File: java/linkedlist/InsertionSortList.java
/**
 * Q16. Insertion Sort List
 * 
 * Problem:
 * Given the head of a singly linked list, sort the list using insertion sort, and return the sorted list's head.
 * 
 * Example:
 * Input: head = [4,2,1,3]
 * Output: [1,2,3,4]
 * 
 * Approach:
 * 1. Create a dummy node to handle the result list
 * 2. For each node in the original list:
 *    - Find the correct position in the sorted list
 *    - Insert the node at that position
 * 3. Return the sorted list
 * 
 * Time Complexity: O(N²) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class InsertionSortList {
    public ListNode insertionSortList(ListNode head) {
        // Handle edge cases
        if (head == null || head.next == null) {
            return head;
        }

        // Create dummy node for the sorted list
        ListNode dummy = new ListNode(0);
        ListNode current = head;

        while (current != null) {
            // Store next node
            ListNode next = current.next;
            
            // Find the correct position in sorted list
            ListNode prev = dummy;
            while (prev.next != null && prev.next.val < current.val) {
                prev = prev.next;
            }
            
            // Insert current node
            current.next = prev.next;
            prev.next = current;
            
            // Move to next node
            current = next;
        }

        return dummy.next;
    }

    // Main method for testing
    public static void main(String[] args) {
        // Create test linked list: 4->2->1->3
        ListNode head = new ListNode(4);
        head.next = new ListNode(2);
        head.next.next = new ListNode(1);
        head.next.next.next = new ListNode(3);

        // Sort the list
        InsertionSortList solution = new InsertionSortList();
        ListNode result = solution.insertionSortList(head);

        // Print the result
        System.out.print("Sorted list using insertion sort: ");
        while (result != null) {
            System.out.print(result.val + " -> ");
            result = result.next;
        }
        System.out.println("null");

        // Test with already sorted list: 1->2->3->4
        ListNode head2 = new ListNode(1);
        head2.next = new ListNode(2);
        head2.next.next = new ListNode(3);
        head2.next.next.next = new ListNode(4);

        ListNode result2 = solution.insertionSortList(head2);
        System.out.print("Already sorted list: ");
        while (result2 != null) {
            System.out.print(result2.val + " -> ");
            result2 = result2.next;
        }
        System.out.println("null");

        // Test with single node: 1
        ListNode head3 = new ListNode(1);
        ListNode result3 = solution.insertionSortList(head3);
        System.out.print("Single node list: ");
        while (result3 != null) {
            System.out.print(result3.val + " -> ");
            result3 = result3.next;
        }
        System.out.println("null");
    }
} 

File: javascript/linkedlist/InsertionSortList.js
/**
 * Insertion Sort List
 * 
 * Problem:
 * Given the head of a singly linked list, sort the list using insertion sort,
 * and return the sorted list's head.
 * 
 * Example:
 * Input: head = [4,2,1,3]
 * Output: [1,2,3,4]
 * Explanation:
 * - Start with 4: [4]
 * - Insert 2: [2,4]
 * - Insert 1: [1,2,4]
 * - Insert 3: [1,2,3,4]
 * 
 * Approach:
 * 1. Create a dummy node to handle edge cases
 * 2. For each node in the original list:
 *    - Find the correct position in the sorted list
 *    - Insert the node at that position
 * 3. Return the sorted list
 * 
 * Time Complexity: O(N²) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class InsertionSortList {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    insertionSortList(head) {
        if (!head || !head.next) return head;

        // Create dummy node for sorted list
        const dummy = new ListNode(0);
        let current = head;

        while (current) {
            // Store next node before we change current's next
            const next = current.next;
            
            // Find the position to insert current node
            let prev = dummy;
            while (prev.next && prev.next.val < current.val) {
                prev = prev.next;
            }

            // Insert current node
            current.next = prev.next;
            prev.next = current;

            // Move to next node
            current = next;
        }

        return dummy.next;
    }

    // Helper method to create a linked list
    static createList(values) {
        if (!values || values.length === 0) return null;
        
        const head = new ListNode(values[0]);
        let current = head;
        
        for (let i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current = current.next;
        }
        
        return head;
    }

    // Helper method to print the list
    static printList(head) {
        let current = head;
        const values = [];
        while (current) {
            values.push(current.val);
            current = current.next;
        }
        console.log(values.join(' -> '));
    }
}

// Test cases
function runTests() {
    const solution = new InsertionSortList();

    // Test case 1: Regular case
    console.log("Test case 1: Regular case");
    const list1 = InsertionSortList.createList([4, 2, 1, 3]);
    console.log("Original list:");
    InsertionSortList.printList(list1);
    console.log("Sorted list:");
    InsertionSortList.printList(solution.insertionSortList(list1));

    // Test case 2: Already sorted list
    console.log("\nTest case 2: Already sorted list");
    const list2 = InsertionSortList.createList([1, 2, 3, 4]);
    console.log("Original list:");
    InsertionSortList.printList(list2);
    console.log("Sorted list:");
    InsertionSortList.printList(solution.insertionSortList(list2));

    // Test case 3: Reverse sorted list
    console.log("\nTest case 3: Reverse sorted list");
    const list3 = InsertionSortList.createList([4, 3, 2, 1]);
    console.log("Original list:");
    InsertionSortList.printList(list3);
    console.log("Sorted list:");
    InsertionSortList.printList(solution.insertionSortList(list3));

    // Test case 4: List with duplicate values
    console.log("\nTest case 4: List with duplicate values");
    const list4 = InsertionSortList.createList([3, 1, 4, 1, 5, 9, 2, 6]);
    console.log("Original list:");
    InsertionSortList.printList(list4);
    console.log("Sorted list:");
    InsertionSortList.printList(solution.insertionSortList(list4));

    // Test case 5: Single node list
    console.log("\nTest case 5: Single node list");
    const list5 = InsertionSortList.createList([1]);
    console.log("Original list:");
    InsertionSortList.printList(list5);
    console.log("Sorted list:");
    InsertionSortList.printList(solution.insertionSortList(list5));

    // Test case 6: Empty list
    console.log("\nTest case 6: Empty list");
    const list6 = null;
    console.log("Original list: null");
    console.log("Sorted list:");
    InsertionSortList.printList(solution.insertionSortList(list6));

    // Test case 7: List with negative numbers
    console.log("\nTest case 7: List with negative numbers");
    const list7 = InsertionSortList.createList([-5, 3, -2, 0, 1, -4]);
    console.log("Original list:");
    InsertionSortList.printList(list7);
    console.log("Sorted list:");
    InsertionSortList.printList(solution.insertionSortList(list7));
}

// Run the tests
runTests(); 